addons/addon-base-raas-appstream/packages/base-raas-appstream-services/lib/appstream/appstream-sc-service.js

```diff
@@ -0,0 +1,207 @@
+    ]);
+
+    // Find stack
+    const { awsAccountId } = await indexesService.mustFind(requestContext, { id: indexId });
+    const {
+      appStreamStackName: stackName,
+      accountId,
+      appStreamFleetName: fleetName,
+    } = await awsAccountsService.mustFind(requestContext, {
+      id: awsAccountId,
+    });
+
+    if (!stackName) {
+      // If stackName is not available then this account does not have AppStream stack configured
+      this.log.warn({
+        msg: `AppStream stack is not configured for AWS Account = ${accountId}. Will not serve the workspace environment via AppStream.`,
+        missingAppStreamConfig: true,
+      });
+      return {};
+    }
+
+    // Verify fleet is associated to appstream stack
+    const appStream = await environmentScService.getClientSdkWithEnvMgmtRole(
+      requestContext,
+      { id: environmentId },
+      { clientName: 'AppStream', options: { signatureVersion: 'v4' } },
+    );
+    const { Names: fleetNames } = await appStream.listAssociatedFleets({ StackName: stackName }).promise();
+
+    if (!_.includes(fleetNames, fleetName)) {
+      throw this.boom.badRequest(`Expected fleet ${fleetName} to be associated with the AppStream stack but found`);
+    }
+
+    return { stackName, fleetName };
+  }
+
+  generateSessionSuffix(environment) {
+    // Generate a unique session suffix for the environment as a 6 character alphanumeric string
+    // This is random looking but string but is deterministic so can be derived from the environment
+    return (new Date(environment.createdAt).getTime() % 36 ** 6).toString(36);
+  }
+
+  generateUserId(requestContext, environment) {
+    // UserId must match [\w+=,.@-]* with max length 32
+    // Don't let the username be too long (otherwise the user won't be able to open multiple sessions)
+    const uid = _.get(requestContext, 'principalIdentifier.uid');
+    // Append a unique session suffix to the user id, this user id is used for creating unique AppStream session
+    // appending suffix to make sure a unique session is created per environment per user
+    const sessionSuffix = this.generateSessionSuffix(environment);
+    return `${uid}-${sessionSuffix}`.replace(/[^\w+=,.@-]+/g, '').slice(0, 32);
+  }
+
+  async urlForFirefox(requestContext, { environmentId, finalDestination }) {
+    const environmentScService = await this.service('environmentScService');
+
+    const appStream = await environmentScService.getClientSdkWithEnvMgmtRole(
+      requestContext,
+      { id: environmentId },
+      { clientName: 'AppStream', options: { signatureVersion: 'v4' } },
+    );
+
+    const environment = await environmentScService.mustFind(requestContext, { id: environmentId });
+
+    const { stackName, fleetName } = await this.getStackAndFleet(requestContext, {
+      environmentId,
+      indexId: environment.indexId,
+    });
+    if (!stackName || !fleetName) {
+      // return original url as is
+      return finalDestination;
+    }
```

Comment:
That's a fair question. Yes, it could be changed to an exception. 

Why I did it this way: I believe we're following the pattern of allowing addons to "contribute" to the final result without needing to check if the feature specific to that addon is enabled.

I'll add a guard that returns the finalDestination if the feature is disabled, and throw an exception if it's enabled and AppStream-resources are not found